diff --git a/src/utilities/onboard.c b/src/utilities/onboard.c
index 8b946f7..f93a1be 100644
--- a/src/utilities/onboard.c
+++ b/src/utilities/onboard.c
@@ -7,6 +7,10 @@
 #include <time.h>
 #include <random.h>
 
+static int is_xemu;
+static const unsigned char xemu_fpga_firmware_id[4] = { 0x58, 0x65, 0x6d, 0x75 }; // 'X','e','m','u' in ASCII
+static const unsigned char xemu_cpld_firmware_id[4] = { 0x4e, 0x6f, 0x77, 0x21 }; // 'N','o','w','!' in ASCII
+
 short i, y;
 unsigned char crt_mode = 1;
 unsigned char last_video_mode = 0;
@@ -352,6 +356,7 @@ void main(void)
   unsigned char idle_time = 0, update_rtc = 0;
 
   mega65_io_enable();
+  is_xemu = !memcmp((void*)0xD632U, xemu_fpga_firmware_id, 4) && !memcmp((void*)0xD62CU, xemu_cpld_firmware_id, 4);
 
   // Disable OSK
   lpoke(0xFFD3615L, 0x7F);
@@ -373,6 +378,8 @@ void main(void)
   // do hardware version dependent setup
   hardware_setup();
 
+restart_screen:
+
   printf("%cWelcome to the MEGA65!\n\n"
          "Before you go further, there are couple of things you need to do.\n\n"
          "Press F3 - F13 to set the time and date (shift toggles direction).", 0x93);
@@ -443,6 +450,32 @@ void main(void)
     POKE(0x286, 1);
     printf(" to save and exit.");
 
+    if (is_xemu) {
+      // Xemu's case, provide own screen, etc ...
+      printf(
+        "%cWelcome to Xemu emulation of MEGA65!\n\n"
+        "This is MEGA65's ONBOARDing utility\n"
+        "running. In case of an emulator it's\n"
+        "not so useful, but still part of the\nprocess.\n\n"
+        "Please press RETURN and accept the\n"
+        "FPGA reconfiguration request of Xemu.\n", 0x93);
+      for (;;) {
+        c = PEEK(0xD610);
+        if (c) POKE(0xD610,0);
+        if (c == 0x0a || c == 0x0d) {
+	   crt_mode = 0;
+	   video_mode = 1;
+	   video_mode_change_pending = 0;
+	   last_video_mode = video_mode;
+           goto write_result;
+	}
+        if (c == 0x4d || c == 0x6d) { // secret 'm' key to allow to run the original utility ...
+           is_xemu = 0;
+           goto restart_screen;
+        }
+      }
+    }
+
     c = PEEK(0xD610);
     if (!c) {
       if (idle_time < 75)
@@ -651,7 +684,7 @@ void main(void)
         last_video_mode = video_mode;
       }
       else {
-
+write_result:
         // First make sure we have the config sector freshly loaded
         lpoke(0xffd3681, 0x01);
         lpoke(0xffd3682, 0x00);
@@ -711,10 +744,16 @@ void main(void)
         while (lpeek(0xffd3680) & 0x03)
           continue;
 
-        printf("%c\n\n\n\n\n\n\n\n\n\n\n"
-               "   Please POWER-CYCLE your MEGA65 now\n"
-               "    by turning it off and on again!", 0x93);
-        while (1);
+	if (!is_xemu) {
+          printf("%c\n\n\n\n\n\n\n\n\n\n\n"
+                 "   Please POWER-CYCLE your MEGA65 now\n"
+                 "    by turning it off and on again!", 0x93);
+          while (1);
+	} else {
+	  printf("\n\nFPGA reconfiguration ...");
+	  while (1)
+            POKE(0xD6CFU,0x42);	// asks for FPGA reconfiguration
+	}
       }
     }
   }
